### PR TITLE
Allow customizing commit message pattern in `prepare-release` command

### LINKED ISSUE #991: New feature - Command "prepare-release"  with a paremeter for add a prefix to the commit's
**Please Describe The Problem To Be Solved:**
In our solution, for every commit generated, we have a format. In this case, we use the "JIRA" card format, for example "LM-xxx-{commit message}". So when the command "prepare-release" is executed, it generates two commits. We need to add this prefix to every commit. So we ask if we can collaborate to add this new feature, which should be an optional parameter to the command.

Actual value : "Set version to '2.1'"
Expected value: "LM-1450 - Set version '2.1'"

### PR DESCRIPTION
**Pull request by fmacavilca**
By default, the `prepare-release` command generates a commit message with the format "Set version to {version}". So with this new option, You can add a prefix or suffix to the default commit message (which will replace just the version).

For example, running the following command on `master`

```
nbgv prepare-release --commit-message-pattern "Custom commit message pattern - {0} custom message"
```

So your commit message is going to be this:

```
Custom commit message pattern '1.0' custom message
```
**Note:** Must include {0} in the pattern to represent the version.

closes #991

### CODE REVIEW

**Comment by fmacavilca** [comment]
@dotnet-policy-service agree

**Review by FernandoCalmet — APPROVED** [review]

**Review by wchoque — APPROVED** [review]

File: doc/nbgv-cli.md [hunk 1]
```diff
@@ -165,6 +165,23 @@ For each branch, the following properties are provided:
 **Note:** When the current branch is already the release branch for the current version, no new branch will be created.
 In that case, the `NewBranch` property will be `null`.
 
+### Customizing the `prepare-release` commit message
```
Comment (AArnott) [inline doc/nbgv-cli.md:168]: Thank you for the documentation update!
It makes me wonder... maybe when the message is customized, the `{0}` should be replaced with only the version itself, allowing total customization of the message.

File: src/nbgv/Program.cs [hunk 2]
```diff
@@ -217,6 +219,7 @@ private static Parser BuildCommandLine()
                 var nextVersion = new Option<string>("--nextVersion", "The version to set for the current branch. If omitted, the next version is determined automatically by incrementing the current version.");
                 var versionIncrement = new Option<string>("--versionIncrement", "Overrides the 'versionIncrement' setting set in version.json for determining the next version of the current branch.");
                 var format = new Option<string>(new[] { "--format", "-f" }, $"The format to write information about the release. Allowed values are: {string.Join(", ", SupportedFormats)}. The default is {DefaultOutputFormat}.").FromAmong(SupportedFormats);
+                var commitMessagePattern = new Option<string>("--commit-message-pattern", "Custom pattern to add a prefix or suffix to the default commit message.");
```
Comment (AArnott) [inline src/nbgv/Program.cs:222]: Can you add to the description something to indicate that `{0}` is expected to appear somewhere in the argument?

File: src/nbgv/Program.cs [hunk 3]
```diff
@@ -763,11 +767,24 @@ private static Task<int> OnPrepareReleaseCommand(string project, string nextVers
                 return Task.FromResult((int)ExitCodes.UnsupportedFormat);
             }
 
+            // validate commit message pattern
+            if (AreCurlyBracesBalanced(commitMessagePattern))
```
Comment (AArnott) [inline src/nbgv/Program.cs:771]: Let's not get into the business of manual validation of the message. Let's leave it to `string.Format` and (if we can) catch any exception thrown in order to print the error message and return the error code.

**Review by AArnott — CHANGES_REQUESTED** [review]
I like it. Just a few changes please. The design change idea (about reducing what `{0}` is replaced with) is just an idea. It's at your discretion.

**Review by AArnott — APPROVED** [review]
Thanks for this!

**Review by FernandoCalmet — APPROVED** [review]

File: src/nbgv/Program.cs [hunk 4]
```diff
@@ -69,6 +70,7 @@ private enum ExitCodes
             ShallowClone,
             InternalError,
             InvalidTagNameSetting,
+            InvalidunformattedCommitMessage,
```
Comment (choquewilber) [inline src/nbgv/Program.cs:73]: @AArnott I think it should be Invalid`U`nformattedCommitMessage

Comment (wchoque) [inline src/nbgv/Program.cs:73]: @AArnott I think it should be Invalid`U`nformattedCommitMessage to follow PascalCase convention

Comment (AArnott) [reply src/nbgv/Program.cs]: Whoops. That was a search and replace result. Yes, I'll fix.

**Comment by fmacavilca** [comment]
@AArnott, do you know when it will be published in the new release?

**Comment by AArnott** [comment]
I can issue a prerelease to nuget.org if that will suffice for you.

**Comment by fmacavilca** [comment]
Yes, please. It will be great. Thanks

**Comment by AArnott** [comment]
https://www.nuget.org/packages/Nerdbank.GitVersioning/3.7.48-alpha

File: doc/nbgv-cli.md (hunk 1 above)
Comment (ronaldbarendse) [reply doc/nbgv-cli.md:168]: Shouldn't the placeholder be named `{version}` instead, both to be more descriptive and align with the release tag and branch name formats?

The current code also uses `string.Format()` without any validation or exception handling, so any unescaped `{...}` values that are not `{0}` will cause a `FormatException` 😢

**Pull request merged**